### PR TITLE
Fix `nng` compilation on aarch64-darwin and update Python and Python dependencies

### PR DESCRIPTION
**Pull request by teymour-aldridge**
- Also make it possible to compile the engine natively on Apple arm64.

## 🔗  Related links

- [Asana Task: Update Python version](https://app.asana.com/0/1199548034582004/1201829883247749/f) (_internal_)
- [Asana Task: Update pyarrow](https://app.asana.com/0/1199548034582004/1202493859582876/f) (_internal_)

## 🌟 What is the purpose of this PR?

Updates Python dependencies in the engine, also makes it possible to run the engine on `aarch64-apple-darwin`

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202493859582876

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -166,7 +166,7 @@ jobs:
         if: matrix.directory == 'packages/engine'
         uses: actions/setup-python@v2
         with:
-          python-version: "3.7"
+          python-version: "3.9"
```
Comment (Alfred-Mountfield) [inline .github/workflows/rust.yml:169]: 🎉 Very happy to see this.

Is there a reason we're not just upgrading to Python 3.10?

File: packages/engine/Makefile.toml [hunk 2]
```diff
@@ -24,7 +24,7 @@ args = ["build", "-p", "memory", "--profile", "${CARGO_MAKE_CARGO_PROFILE}"]
 condition = { env_true = ["CARGO_MAKE_CI"] }
 private = true
 command = "bash"
-args = ["lib/execution/src/runner/python/setup.sh", "python3.7"]
```
Comment (Alfred-Mountfield) [inline packages/engine/Makefile.toml:27]: I can't comment this many lines above but I think you should remove `features build-nng` from above

**Review by Alfred-Mountfield — COMMENTED** [review]
This looks good for the most part (and is very exciting).

When I run them individually, the tests pass locally:
```
 cargo test --package hash_engine_lib --test integration examples::boids_3d::cohesion_arange                             
    Finished test [unoptimized + debuginfo] target(s) in 0.16s
     Running tests/integration.rs (target/debug/deps/integration-74f4d60e53fa4edd)

running 1 test
test examples::boids_3d::cohesion_arange has been running for over 60 seconds
test examples::boids_3d::cohesion_arange ... ok

```

But if I run
```
cargo test --package hash_engine_lib 
```

I get a _**lot**_ of failures, such as:
```
test examples::boids_3d::cohesion_arange ... FAILED
```

where I see an error:
```
ModuleNotFoundError: No module named 'pynng'
```

I'm not entirely sure what's going on there, but you haven't updated the README so I'm not sure what I'm supposed to do 🙃 

I think deleting most of the PR template too quickly might have been a mistake. I'd appreciate at least seeing a section on if this requires changes to the docs (it definitely does and those changes are missing) as well as known issues.

**Comment by TimDiekmann** [comment]
Linked the Asana task

File: packages/engine/Makefile.toml (hunk 2 above)
Comment (TimDiekmann) [reply packages/engine/Makefile.toml:27]: I think generally the build-nng feature should simply be removed

File: packages/engine/README.md [hunk 3]
```diff
@@ -103,41 +101,28 @@ Depending on your needs, different dependencies are required. Building this proj
...
 
 #### For all macs
 
-Unfortunately, Apple currently doesn't provide a way to resize shared-memory allocations. To work around this allocations need to be sufficiently big such that they will not need to be resized. This can be done by setting the `OS_MEMORY_ALLOC_OVERRIDE` environment variable. A reasonable starting value might be `250000000`, but it is heavily dependent on the memory-requirements of your simulation. This can be done using the command line
+Unfortunately, Apple currently doesn't provide a way to resize shared-memory
+allocations. To work around this, allocations need to be sufficiently big such
+that they will not need to be resized. This can be done by setting the
+`OS_MEMORY_ALLOC_OVERRIDE` environment variable. A reasonable starting value
+might be `250000000`, but it is heavily dependent on the memory-requirements of
+your simulation. This can be done using the command line
```
Comment (TimDiekmann) [inline packages/engine/README.md:116]: I think we don't want line breaks per paragraph in Markdown (but using soft line breaks instead in IDEs). I don't really mind either way. cc @kachkaev

File: packages/engine/README.md [hunk 4]
```diff
@@ -103,41 +101,20 @@ Depending on your needs, different dependencies are required. Building this proj
         git checkout $latestTag
         ```
 
-- Python [3.7.x] is required for Python initialization or Python behaviors.
-
+- Python [3.10.x] is required for Python initialization and Python behaviors.
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:104]: Is Python 3.10 _required_? We should probably put _some_ effort into finding out the minimum version we support at the moment.

File: packages/engine/README.md [hunk 5]
```diff
@@ -103,41 +101,20 @@ Depending on your needs, different dependencies are required. Building this proj
...
-- It's necessary to acquire an x86 version of `nng`. Currently, the easiest known way to do this is through:
-
-  - Creating a homebrew installation under Rosetta, [an example guide is here](https://stackoverflow.com/questions/64882584/how-to-run-the-homebrew-installer-under-rosetta-2-on-m1-macbook)
-
-  - Using the x86 brew to install `nng` (which will then install an x86 version). This should result in an nng installation at: `/usr/local/Cellar/nng/1.5.2`
-
-- It's then necessary to set the `NNG_PATH` environment variable
-
-  - The command is likely to be: `export NNG_PATH=/usr/local/Cellar/nng/1.5.2`
+If you want to run Python behaviors, you will need a copy of the open basic linear algebra subroutines (`brew install openblas` - if you do not have Homebrew installed, this is easily done from [its website](https://brew.sh)) - this is currently necessary to install `scipy`. You may also need to install a Fortran compiler (`brew install gfortran`).
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:117]: > You may also need to install a Fortran compiler

🙃 🙃

File: packages/engine/lib/execution/src/runner/python/setup.sh [hunk 6]
```diff
@@ -47,7 +47,28 @@ echo "Running $(python --version) from $(which python)"
 
 echo "Installing python dependencies"
 python -m pip install --upgrade pip
-python -m pip install -r "${SCRIPT_DIR}/requirements.txt"
+if uname -s | grep -q Darwin && uname -p | grep -q arm ; then
+  if ! brew --prefix openblas ; then
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/execution/src/runner/python/setup.sh:51]: I agree that brew is probably the way people should install it, but do we have to require that they do it through brew?

I see we're setting the `PKG_CONFIG_PATH` manually below so maybe for now it's fine to just assume people will use brew and wait until someone encounters a problem and then adjust with new info.. but still not 100% happy with it

File: packages/engine/README.md [hunk 7]
```diff
@@ -103,41 +101,20 @@ Depending on your needs, different dependencies are required. Building this proj
...
         ```
 
-- Python [3.7.x] is required for Python initialization or Python behaviors.
-
+- Python [3.10.x] is required for Python initialization and Python behaviors.
   - Python installation guidance from [their website](https://www.python.org/downloads/)
 
 ### macOS Developer Specific Instructions
 
 #### For all macs
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:109]: I don't _think_ we need this heading any more

File: packages/engine/README.md (hunk 5 above)
Comment (Alfred-Mountfield) [inline packages/engine/README.md:117]: Is this going to be required for ARM-based macs as well?

File: packages/engine/bin/cli/Cargo.toml [hunk 8]
```diff
@@ -20,6 +20,4 @@ tokio = "1.19.2"
 uuid = { version = "1.1.2", features = ["v4", "serde"] }
 
 [features]
-default = ["build-nng"]
```
Comment (Alfred-Mountfield) [inline packages/engine/bin/cli/Cargo.toml:23]: This PR feels like we're doing more than "Updating Python Dependencies" and I'm a little concerned it might be difficult to find in the future. Can we update the PR title, and be careful when clicking merge to update the commit contents too?

**Review by Alfred-Mountfield — COMMENTED** [review]
Thanks @teymour-aldridge! This is such a great QoL change, few last comments.

I've followed the updated README and successfully ran tests on my machine without NNG and such so 👍 from me when the comments are addressed

File: packages/engine/lib/execution/src/runner/python/setup.sh (hunk 6 above)
Comment (Alfred-Mountfield) [reply packages/engine/lib/execution/src/runner/python/setup.sh:51]: (general comment, we don't have to block merging on this)

File: packages/engine/README.md (hunk 5 above)
Comment (Alfred-Mountfield) [reply packages/engine/README.md:117]: As it's currently under the heading for all macs

Comment (Alfred-Mountfield) [reply packages/engine/README.md:117]: Out of curiosity, how come?

Comment (Alfred-Mountfield) [reply packages/engine/README.md:117]: What's different on Mac vs Linux re this

File: packages/engine/README.md (hunk 4 above)
Comment (TimDiekmann) [reply packages/engine/README.md:104]: I agree but (without checking) I guess, that Python 3.10 is not distributed on all flavors. I guess we can say, that we target Python 3.10 but it *may* also work with older Python versions.

Comment (TimDiekmann) [reply packages/engine/README.md:104]: Maybe we should add a `Python` field on https://github.com/hashintel/hash/blob/main/.github/ISSUE_TEMPLATE/bug-report-hash-engine.md

File: packages/engine/README.md [hunk 9]
```diff
@@ -101,13 +101,12 @@ Depending on your needs, different dependencies are required. Building this proj
         git checkout $latestTag
         ```
 
-- Python [3.10.x] is required for Python initialization and Python behaviors.
+- Python [3.7.x] is required for Python initialization or Python behaviors.
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:104]: Can we rephrase this to 
```suggestion
- Python [3.x >= 3.7] is required for Python initialization or Python behaviors.
```

File: packages/engine/README.md (hunk 5 above)
Comment (Alfred-Mountfield) [reply packages/engine/README.md:116]: If these aren't Mac specific then can we move these to a different section. If there isn't an obvious space perhaps we could have a "potential dependencies" or "troubleshooting" or something where we list these (I think there was a place where we say "depending on how lightweight you install is" for example)

**Review by Alfred-Mountfield — APPROVED** [review]
Nice! Might be worth pinging @TimDiekmann quickly to see how strongly he feels about comment chain he's got open, but looks good from my side

**Comment by TimDiekmann** [comment]
One comment was addressed, the other comment can be addressed when revisiting the issue templates, so I'm not blocking the PR to be merged. 🚀
